executor.go

```diff
@@ -376,6 +397,105 @@ func containerIsPresent(containers []docker.APIContainers, containerId string) b
 	return false
 }
 
+// maybePullContainer checks if we need to pull a container and does so if needed.
+func (exec *sidecarExecutor) maybePullContainer(taskInfo *mesos.TaskInfo) error {
+	var shouldPullContainer bool
+
+	if !container.CheckImage(exec.client, taskInfo) {
+		shouldPullContainer = true
+	}
+
+	if taskInfo.Container.Docker.ForcePullImage != nil && *taskInfo.Container.Docker.ForcePullImage {
+		shouldPullContainer = true
+	}
+
+	log.Infof("Using image '%s'", taskInfo.Container.Docker.Image)
+
+	// Pull the image if it's stale/missing or we're told to force it
+	if shouldPullContainer {
+		err := container.PullImage(exec.client, taskInfo, exec.dockerAuth)
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Info("Re-using existing image... already present")
+
+	return nil
+}
+
+// monitorAWSCredsLease will be run in a background goroutine and will shut down the managed
+// process if we are about to hit our expiry. We don't bother with expiring the lease here,
+// it will be handled when the looper shuts down. If that somehow fails, it will still get
+// cleaned up by Vault afterward.
+func (exec *sidecarExecutor) monitorAWSCredsLease() {
+	wait := func() {
+		vars := exec.awsCredsLease
+
+		log.Infof("Monitoring AWS Credentials expiry at '%s' for lease ID '%s'", vars.LeaseExpiryTime, vars.LeaseID)
+
+		// Block until expiry
+		<-time.After(vars.LeaseExpiryTime.Sub(time.Now().UTC()))
+	}
+
+	// We may end up in here more than once if the lease is renewed. This will reset us to
+	// monitor the new lease timeout.
+	for exec.awsCredsLease.LeaseExpiryTime.After(time.Now().UTC()) {
+		wait()
+	}
+
+	// Send *ourselves* a TERM to not have yet another way to shut down
```

Comment:
This is a good reminder that shutdown behavior is kind of nasty in this thing. A future refactor should clean this up. For now sending ourselves a signal is the cleanest way to propagate the message to shut down.